dask/array/core.py

```diff
@@ -1497,6 +1497,16 @@ def __xor__(self, other):
     def __rxor__(self, other):
         return elemwise(operator.xor, other, self)
 
+    def __matmul__(self, other):
+        return tensordot(self, other, axes=((self.ndim - 1,),
+                                            (other.ndim - 2,)))
+
+    def __rmatmul__(self, other):
+        if not hasattr(other, 'ndim'):
+            other = np.asarray(other)  # account for array-like on LHS
+        return tensordot(other, self, axes=((other.ndim - 1,),
+                                            (self.ndim - 2,)))
```

Comment:
@mrocklin Whew, made it through that discussion. It looks like you collectively punted on this topic there. However, the main reason was that you didn't want to assume a dense array in the context of sparseness. imho dask.array is in a different category, because it is designed to mimic dense NumPy arrays.

Let's wait for @shoyer's regarding @ vs dot, at which point I'll move this casting to the dot function where it makes more sense (as suggest by @jakevdp).